pkg/server/server.go

```diff
@@ -0,0 +1,122 @@
+package server
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/wzshiming/s3d/pkg/auth"
+	"github.com/wzshiming/s3d/pkg/storage"
+)
+
+// S3Server represents the S3-compatible server
+type S3Server struct {
+	storage       *storage.Storage
+	authenticator *auth.Authenticator
+}
+
+// NewS3Server creates a new S3 server
+func NewS3Server(storage *storage.Storage, authenticator *auth.Authenticator) *S3Server {
+	return &S3Server{
+		storage:       storage,
+		authenticator: authenticator,
+	}
+}
+
+// Handler returns the HTTP handler
+func (s *S3Server) Handler() http.Handler {
+	mux := http.NewServeMux()
+
+	// Wrap with auth middleware
+	handler := s.authenticator.AuthMiddleware(http.HandlerFunc(s.handleRequest))
+	mux.Handle("/", handler)
+
+	return mux
+}
+
+// handleRequest handles all S3 requests
+func (s *S3Server) handleRequest(w http.ResponseWriter, r *http.Request) {
```

Comment:
cyclomatic complexity 24 of func `(*S3Server).handleRequest` is high (> 15) (gocyclo)